File: python/tutorials/tour-of-orchestration-python/app/getstarted/service.py
import restate

from app.utils import create_recurring_payment, create_subscription
from app.types import SubscriptionRequest


subscription_service = restate.Service("SubscriptionService")


@subscription_service.handler()
async def add(ctx: restate.Context, req: SubscriptionRequest) -> None:
    payment_id = str(ctx.uuid())

    pay_ref = await ctx.run_typed(
        "pay",
        create_recurring_payment,
        credit_card=req.credit_card,
        payment_id=payment_id,
    )

    for subscription in req.subscriptions:
        await ctx.run_typed(
            f"add-{subscription}",
            create_subscription,
            user_id=req.user_id,
            subscription=subscription,
            payment_ref=pay_ref,
        )

File: kotlin/tutorials/tour-of-orchestration-kotlin/src/main/kotlin/my/example/getstarted/service.kt
import dev.restate.sdk.Context
import dev.restate.sdk.annotation.Handler
import dev.restate.sdk.annotation.Service
import dev.restate.sdk.http.vertx.RestateHttpServer
import dev.restate.sdk.kotlin.endpoint
import java.util.*

data class SubscriptionRequest(
    val userId: String,
    val creditCard: String,
    val subscriptions: Array<String>
)

// Mock function to create recurring payment
suspend fun createRecurringPayment(creditCard: String, paymentId: String): String {
    return "payRef-${UUID.randomUUID()}"
}

// Mock function to create subscription
suspend fun createSubscription(userId: String, subscription: String, paymentRef: String) {
    println("Creating subscription for user: $userId, subscription: $subscription, paymentRef: $paymentRef")
}

@Service
class SubscriptionService {

    @Handler
    suspend fun add(ctx: Context, req: SubscriptionRequest) {
        val paymentId = ctx.random().nextUUID().toString()

        val payRef = ctx.runBlock(
            "pay",
            String::class
        ) {
            createRecurringPayment(req.creditCard, paymentId)
        }

        for (subscription in req.subscriptions) {
            ctx.runBlock("add-$subscription") {
                createSubscription(req.userId, subscription, payRef)
            }
        }
    }
}

fun main() {
    RestateHttpServer.listen(endpoint {
        bind(SubscriptionService())
    })
}